File: frontend/src/lib/logger.ts
import winston from 'winston';
import 'winston-daily-rotate-file';
import path from 'path';

// Define log directory (outside src source)
// In Docker, we usually map this volume
const LOG_DIR = path.join(process.cwd(), 'logs');

// Define log format
const logFormat = winston.format.combine(
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
    winston.format.printf(({ timestamp, level, message, ...meta }) => {
        return `${timestamp} [${level.toUpperCase()}]: ${message} ${Object.keys(meta).length ? JSON.stringify(meta) : ''
            }`;
    })
);

// Create logger instance
const logger = winston.createLogger({
    level: process.env.NODE_ENV === 'development' ? 'debug' : 'info',
    format: winston.format.json(), // Base format for files is often JSON
    transports: [
        // 1. Console transport (for docker logs)
        new winston.transports.Console({
            format: winston.format.combine(
                winston.format.colorize(),
                logFormat
            ),
        }),
    ],
});

// Add file transports ONLY on the server side
if (typeof window === 'undefined') {
    // Daily rotate file for all logs
    logger.add(
        new winston.transports.DailyRotateFile({
            filename: path.join(LOG_DIR, 'app-%DATE%.log'),
            datePattern: 'YYYY-MM-DD',
            zippedArchive: true,
            maxSize: '20m',
            maxFiles: '14d',
            format: winston.format.combine(winston.format.timestamp(), winston.format.json()),
        })
    );

    // Separate file for errors
    logger.add(
        new winston.transports.DailyRotateFile({
            filename: path.join(LOG_DIR, 'error-%DATE%.log'),
            datePattern: 'YYYY-MM-DD',
            zippedArchive: true,
            maxSize: '20m',
            maxFiles: '30d',
            level: 'error',
            format: winston.format.combine(winston.format.timestamp(), winston.format.json()),
        })
    );
}

export default logger;

File: backend/utils/logger.py
import sys
from loguru import logger
from config import settings
import os

# Создаем папку для логов, если её нет
LOG_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "logs")
os.makedirs(LOG_DIR, exist_ok=True)

def setup_logger():
    """
    Настройка глобального логгера Loguru.
    - JSON формат для файлов
    - Ротация каждый день
    - Разделение уровней (INFO, ERROR)
    """
    # Удаляем стандартный обработчик, чтобы не дублировать
    logger.remove()

    # 1. Логи в консоль (для разработки удобнее цветной текст, а не JSON)
    logger.add(
        sys.stdout,
        format="<green>{time:YYYY-MM-DD HH:mm:ss}</green> | <level>{level: <8}</level> | <cyan>{name}</cyan>:<cyan>{function}</cyan>:<cyan>{line}</cyan> - <level>{message}</level>",
        level="INFO" if not settings.debug else "DEBUG",
        colorize=True
    )

    # 2. Логи в файл (JSON, все уровни, ротация ежедневно)
    # Используем serialize=True для сохранения в JSON
    logger.add(
        os.path.join(LOG_DIR, "app_{time:YYYY-MM-DD}.json"),
        rotation="00:00",  # Новый файл каждый день в полночь
        retention="14 days",  # Хранить 14 дней
        compression="zip",  # Сжимать старые логи
        serialize=True,     # Писать в формате JSON
        level="INFO",
        encoding="utf-8"
    )

    # 3. Отдельный файл только для ошибок (чтобы быстро находить проблемы)
    logger.add(
        os.path.join(LOG_DIR, "errors.log"),
        rotation="10 MB",
        retention="30 days",
        level="ERROR",
        format="{time:YYYY-MM-DD HH:mm:ss} | {level} | {name}:{function}:{line} - {message}",
        encoding="utf-8" # Для ошибок JSON может быть избыточен, лучше читаемый текст, но можно и JSON
    )

    return logger

# Инициализируем при импорте, но функцию setup_logger можно вызвать повторно для реконфигурации
logger = setup_logger()
